transport-timeout/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Wraps around a `Transport` and adds a timeout to all the incoming and outgoing connections.
+//!
+//! The timeout includes the upgrading process.
+// TODO: add example
+
+#[macro_use]
+extern crate futures;
+extern crate libp2p_core;
+#[macro_use]
+extern crate log;
+extern crate tokio_timer;
+
+use std::io::{Error as IoError, ErrorKind as IoErrorKind};
+use std::time::{Duration, Instant};
+use futures::{Future, Poll, Async, Stream};
+use libp2p_core::{Transport, Multiaddr, MuxedTransport};
+use tokio_timer::{Deadline, DeadlineError};
+
+/// Wraps around a `Transport` and adds a timeout to all the incoming and outgoing connections.
+///
+/// The timeout includes the upgrade. There is no timeout on the listener or on stream of incoming
+/// substreams.
+#[derive(Debug, Copy, Clone)]
+pub struct TransportTimeout<InnerTrans> {
+    inner: InnerTrans,
+    outgoing_timeout: Duration,
+    incoming_timeout: Duration,
+}
+
+impl<InnerTrans> TransportTimeout<InnerTrans> {
+    /// Wraps around a `Transport` to add timeouts to all the sockets created by it.
+    #[inline]
+    pub fn new(trans: InnerTrans, timeout: Duration) -> Self {
+        TransportTimeout {
+            inner: trans,
+            outgoing_timeout: timeout,
+            incoming_timeout: timeout,
```

Comment:
Since there is support for different timeouts on incoming and outgoing connections, maybe allow client code to pass in two separate `Duration`s?